website/blog/2024-03-18-deep-learning-part-2/index.md

```diff
@@ -0,0 +1,230 @@
+
+Overall, fine-tuning Llama2-7B with PEFT is a promising approach for instruction following tasks and similar applications that require working with large models on limited resources.
+
+
+
+# Setting Up MLflow for Fine tuning
+
+The code provided below illustrates the setup for Fine tuning, augmented with MLflow integration. We have marked the sections used for MLflow tracking. At the beginning of the functions, MLflow is configured, initiating a run that automatically logs the experiment. The details of  `callback.py` will be discussed in a later section.
+
+This example demonstrates how to set up a fine tuning job on Databricks. A similar setup can be achieved on any platform capable of supporting this type of workload.
+
+We are using the [e2e_nlg](https://huggingface.co/datasets/e2e_nlg) dataset as it is a great dataset is used for training end-to-end, data-driven natural language generation systems in the restaurant domain
+
+```bash
+!export DATABRICKS_TOKEN && export DATABRICKS_HOST && export MLFLOW_EXPERIMENT_NAME && export MLFLOW_FLATTEN_PARAMS &&  \
+ export HF_MLFLOW_LOG_ARTIFACTS && export MLFLOW_ENABLE_SYSTEM_METRICS_LOGGING && cd .. && deepspeed \
+--num_gpus='{num_gpus}' \
+--module databricks_llm.fine_tune \
+--final_model_output_path='{dbfs_output_location}' \
+--output_dir="/local_disk0/output" \
+--dataset={dataset} \
+--model={pretrained_name_or_path} \
+--tokenizer={pretrained_name_or_path} \
+--use_lora=false \
+--use_4bit=false \
+--deepspeed_config="ds_configs/ds_zero_3_cpu_offloading.json" \
+--fp16=false \
+--bf16=true \
+--per_device_train_batch_size=16 \
+--per_device_eval_batch_size=48 \
+--gradient_checkpointing=true \
+--gradient_accumulation_steps=1 \
+--learning_rate=1e-6 \
+--lr_scheduler_type="cosine" \
+--warmup_steps=50 \
+--evaluation_strategy="steps" \
+--save_strategy="steps" \
+--save_steps=100 \
+--num_train_epochs=3\
+--run_name="{pretrained_name_or_path}-blog-run"
+``````
+See code details [here](https://github.com/puneet-jain159/fine_tuning_llm_with_mlflow_blog/blob/495c156f73435996706c42bfda4bb7519dc21dab/notebooks/run_deepspeed_fine_tuning.py#L117)
+
+In the above code block we are doing the following: Logging system metrics with MLFlow ensures that CPU stats, GPU stats, memory usage, network traffic, and disk usage during the training process is logged. Particularly, the ability to log and track GPU utilization during a training run is critical and thus being able to log it intermittently is particularly useful. The training run is parametrized such that the model is trained for 3 epochs, with checkpoints saved and metrics logged every 100 training steps. The run is given an appropriate name based on the path of the pretrained model.
+
+The diagram below is a visualization of how finetuning and MLflow fit together.
+
+![](finetune_with_MLflow.png)
+
+Using MLflow while fine tuning ML model in Databricks
+
+
+# The Importance of Logging
```

Comment:
H1 is for the title of the blog; please use h2 for major section headers and h3 for subsection headers